### DISCUSSION TITLE
Deepen headings one level for selected text

### DISCUSSION THREAD

**Discussion by carlosbaraza** [SilentVoid13/Templater#605] [Templates Showcase]
Input
```
# Heading
## Subheading
```

Output:
```
## Heading
### Subheading
```

## Template
```
<%*
let text = await tp.file.selection()
	.split('\n')
	.map(t => t.match(/^[#]+\s+/gi) ? `#${t}` : t)
	.join('\n');
%><% text %>
```

**Comment by ManasMadrecha** [comment]
@carlos Amazing, works like a charm. What would be the code to do the reverse, i.e. increase the Heading levels for selected text?

**Reply by ManasMadrecha** [reply]
Got it! Just remove the 1st character of the line, i.e. `#`.
```js
<%* let text = await tp.file.selection().split('\n').map(t => t.match(/^[#]+\s+/gi) ? t.substring(1) : t).join('\n'); %>
<% text %>
```

**Comment by MarredCheese** [comment]
### Opportunities for improvement
1. Multiple versions for either increasing or decreasing header depth
2. Multiple versions for operating on either a selection or all preceding text
3. You're not guarding against the invalid target of some `#` followed by some space, with no words afterward.
4. You're not guarding against invalid results, like changing header 6 to header 7.
5. You don't need to split, map, and join.  Just use regex replaceAll.
6. The regex doesn't need the `i` flag or brackets around the `#`.

### Result

#### Increase Header Depth

```js
<%*
// This code will insert a modified copy of the selection.  Afterward, you'll
// need to delete the original version of the selection manually.
tR += await tp.file.selection().replaceAll(/^(?=#{1,5}[^\S\r\n]+\S)/gm, '#')
%>

<%*
// This code will update everything before it.  Put it at the end of the
// file to update the whole note.
tR = tR.replaceAll(/^(?=#{1,5}[^\S\r\n]+\S)/gm, '#')
%>
```

#### Decrease Header Depth

```js
<%*
// This code will insert a modified copy of the selection.  Afterward, you'll
// need to delete the original version of the selection manually.
tR += await tp.file.selection().replaceAll(/^#(?=#{1,5}[^\S\r\n]+\S)/gm, '')
%>

<%*
// This code will update everything before it.  Put it at the end of the
// file to update the whole note.
tR = tR.replaceAll(/^#(?=#{1,5}[^\S\r\n]+\S)/gm, '')
%>
```

**Reply by ManasMadrecha** [reply]
Much more elegant 👌🏻
